Upgrade from older versions fails with "no such table: metadata"

upgrading from 250.241025.0

```
running offline db migration from 1 to 5
Logger.ts:89 migration finished
Logger.ts:99 Error while initializing offline cache storage Error: sqlite error(1): no such table: metadata
    at new Kt (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:20582)
    at Gt.prepare (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:22431)
    at b.run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:6861)
    at run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:7487)
    at Bt.handleMessage (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19204)
    at file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:18285
    at MessagePort.<anonymous> (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19794)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
error @ Logger.ts:99
Logger.ts:89 error Error: sqlite error(1): no such table: metadata
    at new Kt (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:20582)
    at Gt.prepare (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:22431)
    at b.run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:6861)
    at run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:7487)
    at Bt.handleMessage (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19204)
    at file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:18285
    at MessagePort.<anonymous> (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19794)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28) Error: sqlite error(1): no such table: metadata
    at new Kt (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:20582)
    at Gt.prepare (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:22431)
    at b.run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:6861)
    at run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:7487)
    at Bt.handleMessage (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19204)
    at file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:18285
    at MessagePort.<anonymous> (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19794)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
Logger.ts:89 Unknown error Error: sqlite error(1): no such table: metadata
    at new Kt (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:20582)
    at Gt.prepare (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:22431)
    at b.run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:6861)
    at run (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:7487)
    at Bt.handleMessage (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19204)
    at file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:18285
    at MessagePort.<anonymous> (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19794)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
Logger.ts:89 ignoring error to read typeModel from filesystem: Error: ENOENT: no such file or directory, open '/home/ivk/.config/tutanota-desktop-test/server_type_models.json'
Logger.ts:89 re-initializing server model from new server response data
Logger.ts:89 ws connect reconnect: false state: terminated
Logger.ts:89 initIndexer after log in
Logger.ts:89 send indexer error to main thread Error: sqlite error(1): no such table: search_metadata
    at new Kt (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:20582)
    at Gt.prepare (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:22431)
    at b.get (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:6917)
    at get (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/sqlworker.js:1:7413)
    at Bt.handleMessage (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19204)
    at file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:18285
    at MessagePort.<anonymous> (file:///tmp/.mount_tutanohUnKn4/resources/app.asar/desktop/node-sqlcipher-DlVGlcMu.js:1:19794)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:827:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)
```

--- a/src/common/api/worker/offline/OfflineStorage.ts
+++ b/src/common/api/worker/offline/OfflineStorage.ts
@@ -25,7 +25,6 @@ import {
 	parseTypeString,
 	splitInChunks,
 	typedEntries,
-	typedKeys,
 	typedValues,
 	TypeRef,
 } from "@tutao/tutanota-utils"
@@ -36,7 +35,7 @@ import { OfflineStorageMigrator } from "./OfflineStorageMigrator.js"
 import { CustomCacheHandlerMap } from "../rest/cacheHandler/CustomCacheHandler.js"
 import { InterWindowEventFacadeSendDispatcher } from "../../../native/common/generatedipc/InterWindowEventFacadeSendDispatcher.js"
 import { SqlCipherFacade } from "../../../native/common/generatedipc/SqlCipherFacade.js"
-import { FormattedQuery, SqlValue, TaggedSqlValue, untagSqlObject } from "./SqlValue.js"
+import { FormattedQuery, SqlValue, TaggedSqlValue, untagSqlObject, untagSqlValue } from "./SqlValue.js"
 import { Type as TypeId } from "../../common/EntityConstants.js"
 import { OutOfSyncError } from "../../common/error/OutOfSyncError.js"
 import { sql, SqlFragment } from "./Sql.js"
@@ -84,16 +83,41 @@ export interface OfflineDbMeta {
 
 export const TableDefinitions = Object.freeze({
 	// plus ownerGroup added in a migration
-	list_entities:
-		"type TEXT NOT NULL, listId TEXT NOT NULL, elementId TEXT NOT NULL, ownerGroup TEXT, entity BLOB NOT NULL, PRIMARY KEY (type, listId, elementId)",
+	list_entities: {
+		definition:
+			"CREATE TABLE IF NOT EXISTS list_entities (type TEXT NOT NULL, listId TEXT NOT NULL, elementId TEXT NOT NULL, ownerGroup TEXT, entity BLOB NOT NULL, PRIMARY KEY (type, listId, elementId))",
+		purgedWithCache: true,
+	},
 	// plus ownerGroup added in a migration
-	element_entities: "type TEXT NOT NULL, elementId TEXT NOT NULL, ownerGroup TEXT, entity BLOB NOT NULL, PRIMARY KEY (type, elementId)",
-	ranges: "type TEXT NOT NULL, listId TEXT NOT NULL, lower TEXT NOT NULL, upper TEXT NOT NULL, PRIMARY KEY (type, listId)",
-	lastUpdateBatchIdPerGroupId: "groupId TEXT NOT NULL, batchId TEXT NOT NULL, PRIMARY KEY (groupId)",
-	metadata: "key TEXT NOT NULL, value BLOB, PRIMARY KEY (key)",
-	blob_element_entities:
-		"type TEXT NOT NULL, listId TEXT NOT NULL, elementId TEXT NOT NULL, ownerGroup TEXT, entity BLOB NOT NULL, PRIMARY KEY (type, listId, elementId)",
-} as const)
+	element_entities: {
+		definition:
+			"CREATE TABLE IF NOT EXISTS element_entities (type TEXT NOT NULL, elementId TEXT NOT NULL, ownerGroup TEXT, entity BLOB NOT NULL, PRIMARY KEY (type, elementId))",
+		purgedWithCache: true,
+	},
+	ranges: {
+		definition:
+			"CREATE TABLE IF NOT EXISTS ranges (type TEXT NOT NULL, listId TEXT NOT NULL, lower TEXT NOT NULL, upper TEXT NOT NULL, PRIMARY KEY (type, listId))",
+		purgedWithCache: true,
+	},
+	lastUpdateBatchIdPerGroupId: {
+		definition: "CREATE TABLE IF NOT EXISTS lastUpdateBatchIdPerGroupId (groupId TEXT NOT NULL, batchId TEXT NOT NULL, PRIMARY KEY (groupId))",
+		purgedWithCache: true,
+	},
+	metadata: {
+		definition: "CREATE TABLE IF NOT EXISTS metadata (key TEXT NOT NULL, value BLOB, PRIMARY KEY (key))",
+		purgedWithCache: false,
+		onBeforePurged: async (sqlCipherFacade: SqlCipherFacade) => {
+			if (await tableExists(sqlCipherFacade, "metadata")) {
+				await sqlCipherFacade.run("DELETE FROM metadata WHERE key = 'lastUpdateTime'", [])
+			}
+		},
+	},
+	blob_element_entities: {
+		definition:
+			"CREATE TABLE IF NOT EXISTS blob_element_entities (type TEXT NOT NULL, listId TEXT NOT NULL, elementId TEXT NOT NULL, ownerGroup TEXT, entity BLOB NOT NULL, PRIMARY KEY (type, listId, elementId))",
+		purgedWithCache: true,
+	},
+} as const) satisfies Record<string, OfflineStorageTable>
 
 type Range = { lower: Id; upper: Id }
 
@@ -130,12 +154,20 @@ export interface OfflineStorageTable {
 	 * deleted.
 	 */
 	purgedWithCache: boolean
+
+	/**
+	 * Action to perform before the table is dropped
+	 *
+	 * Could also be used to perform an action **instead** of dropping the table when {@link purgedWithStorage} is false
+	 */
+	onBeforePurged?: (sqlCipherFacade: SqlCipherFacade) => Promise<void>
 }
 
 export class OfflineStorage implements CacheStorage {
 	private userId: Id | null = null
 	private databaseKey: Uint8Array | null = null
 	private timeRangeDate: Date | null = null
+	private readonly allTables: Record<string, OfflineStorageTable>
 
 	constructor(
 		private readonly sqlCipherFacade: SqlCipherFacade,
@@ -146,9 +178,10 @@ export class OfflineStorage implements CacheStorage {
 		private readonly modelMapper: ModelMapper,
 		private readonly typeModelResolver: TypeModelResolver,
 		private readonly customCacheHandler: CustomCacheHandlerMap,
-		private readonly additionalTables: Record<string, OfflineStorageTable>,
+		additionalTables: Record<string, OfflineStorageTable>,
 	) {
 		assert(isOfflineStorageAvailable() || isTest(), "Offline storage is not available.")
+		this.allTables = Object.freeze(Object.assign({}, additionalTables, TableDefinitions))
 	}
 
 	async getWholeListParsed(typeRef: TypeRef<unknown>, listId: string): Promise<ServerModelParsedInstance[]> {
@@ -203,6 +236,9 @@ export class OfflineStorage implements CacheStorage {
 				throw e
 			}
 		}
+		// We createTables again in case they were purged in a migration
+		await this.createTables()
+
 		// if nothing is written here, it means it's a new database
 		return (await this.getLastUpdateTime()).type === "never"
 	}
@@ -563,11 +599,10 @@ export class OfflineStorage implements CacheStorage {
 			return
 		}
 
-		for (const tableName of typedKeys(TableDefinitions)) {
-			await this.sqlCipherFacade.run(`DROP TABLE IF EXISTS ${tableName}`, [])
-		}
-
-		for (const [tableName, { purgedWithCache }] of typedEntries(this.additionalTables)) {
+		for (const [tableName, { purgedWithCache, onBeforePurged }] of typedEntries(this.allTables)) {
+			if (onBeforePurged != null) {
+				await onBeforePurged(this.sqlCipherFacade)
+			}
 			if (purgedWithCache) {
 				await this.sqlCipherFacade.run(`DROP TABLE IF EXISTS ${tableName}`, [])
 			}
@@ -732,16 +767,7 @@ export class OfflineStorage implements CacheStorage {
 	}
 
 	private async createTables() {
-		for (let [name, definition] of typedEntries(TableDefinitions)) {
-			await this.sqlCipherFacade.run(
-				`CREATE TABLE IF NOT EXISTS ${name}
-                 (
-                     ${definition}
-                 )`,
-				[],
-			)
-		}
-		for (const { definition } of typedValues(this.additionalTables)) {
+		for (const { definition } of typedValues(this.allTables)) {
 			await this.sqlCipherFacade.run(definition, [])
 		}
 	}
@@ -973,4 +999,13 @@ export interface OfflineStorageCleaner {
 	 * Delete instances from db that are older than timeRangeDays.
 	 */
 	cleanOfflineDb(offlineStorage: OfflineStorage, timeRangeDate: Date | null, userId: Id, now: number): Promise<void>
+}
+
+export async function tableExists(sqlCipherFacade: SqlCipherFacade, table: string): Promise<boolean> {
+	// Read the schema for the table https://sqlite.org/schematab.html
+	const { query, params } = sql`SELECT COUNT(*) as metadata_exists
+                                FROM sqlite_schema
+                                WHERE name = ${table}`
+	const result = assertNotNull(await sqlCipherFacade.get(query, params))
+	return untagSqlValue(result["metadata_exists"]) === 1
 }

--- a/test/tests/api/worker/offline/OfflineStorageTest.ts
+++ b/test/tests/api/worker/offline/OfflineStorageTest.ts
@@ -49,7 +49,7 @@ import {
 } from "../../../../../src/common/api/entities/tutanota/TypeRefs.js"
 import { OfflineStorageMigrator } from "../../../../../src/common/api/worker/offline/OfflineStorageMigrator.js"
 import { InterWindowEventFacadeSendDispatcher } from "../../../../../src/common/native/common/generatedipc/InterWindowEventFacadeSendDispatcher.js"
-import { untagSqlObject } from "../../../../../src/common/api/worker/offline/SqlValue.js"
+import { SqlType, untagSqlObject } from "../../../../../src/common/api/worker/offline/SqlValue.js"
 import { MailSetKind } from "../../../../../src/common/api/common/TutanotaConstants.js"
 import { Type as TypeId } from "../../../../../src/common/api/common/EntityConstants.js"
 import { GroupMembershipTypeRef, User, UserTypeRef } from "../../../../../src/common/api/entities/sys/TypeRefs.js"
@@ -151,14 +151,16 @@ o.spec("OfflineStorageDb", function () {
 	}
 
 	o.spec("additionalTables", () => {
+		let sqlMock: SqlCipherFacade
+
 		o.beforeEach(async () => {
+			sqlMock = object()
 			// to satisfy the external o.afterEach()
 			// we won't actually use this storage instance for these tests, since we don't want to test with real facades
 			await storage.init({ userId, databaseKey, timeRangeDate, forceNewDatabase: false })
 		})
 
 		o.test("init calls createTables which initializes all tables", async () => {
-			const sqlMock: SqlCipherFacade = object()
 			const storageWithMockedSql = new OfflineStorage(sqlMock, object(), object(), object(), object(), object(), object(), object(), {
 				some_table: {
 					definition: "some statement will be run here",
@@ -184,7 +186,10 @@ o.spec("OfflineStorageDb", function () {
 		})
 
 		o.test("purgeStorage purges all purgeable tables", async () => {
-			const sqlMock: SqlCipherFacade = object()
+			when(sqlMock.get(matchers.contains("SELECT COUNT(*) as metadata_exists"), matchers.anything())).thenResolve({
+				metadata_exists: { type: SqlType.Number, value: 1 },
+			})
+
 			const storageWithMockedSql = new OfflineStorage(sqlMock, object(), object(), object(), object(), object(), object(), object(), {
 				some_table: {
 					definition: "some statement will be run here",
@@ -201,8 +206,52 @@ o.spec("OfflineStorageDb", function () {
 			verify(sqlMock.run("DROP TABLE IF EXISTS some_table", []), { times: 0 })
 
 			for (const table of typedKeys(TableDefinitions)) {
-				verify(sqlMock.run(`DROP TABLE IF EXISTS ${table}`, []))
+				verify(sqlMock.run(`DROP TABLE IF EXISTS ${table}`, []), {
+					times: TableDefinitions[table].purgedWithCache ? 1 : 0,
+				})
 			}
+		})
+
+		o.test("purgeStorage calls onBeforePurged", async () => {
+			when(sqlMock.get(matchers.contains("SELECT COUNT(*) as metadata_exists"), matchers.anything())).thenResolve({
+				metadata_exists: { type: SqlType.Number, value: 1 },
+			})
+
+			const storageWithMockedSql = new OfflineStorage(sqlMock, object(), object(), object(), object(), object(), object(), object(), {
+				some_table: {
+					definition: "some statement will be run here",
+					purgedWithCache: false,
+					async onBeforePurged(sqlCipherFacade: SqlCipherFacade) {
+						sqlCipherFacade.run("onBeforePurged was called", [])
+					},
+				},
+				another_table: {
+					definition: "another statement will be run here",
+					purgedWithCache: true,
+				},
+			})
+			await storageWithMockedSql.init({ userId, databaseKey, timeRangeDate, forceNewDatabase: false })
+			await storageWithMockedSql.purgeStorage()
+			verify(sqlMock.run("onBeforePurged was called", []), { times: 1 })
+		})
+
+		o.test("tables are created after migration", async function () {
+			when(sqlMock.get(matchers.contains("SELECT COUNT(*) as metadata_exists"), matchers.anything())).thenResolve({
+				metadata_exists: { type: SqlType.Number, value: 1 },
+			})
+
+			const storageWithMockedSql = new OfflineStorage(sqlMock, object(), object(), object(), object(), object(), object(), object(), {
+				some_table: {
+					definition: "some statement will be run here",
+					purgedWithCache: false,
+				},
+			})
+
+			when(migratorMock.migrate(storageWithMockedSql, dbFacade)).thenDo(() => {
+				verify(sqlMock.run("some statement will be run here", []), { times: 1 })
+			})
+			await storageWithMockedSql.init({ userId, databaseKey, timeRangeDate, forceNewDatabase: false })
+			verify(sqlMock.run("some statement will be run here", []), { times: 2 })
 		})
 	})
 

